diff --git a/src/client.c b/src/client.c
index 15bb44f..e4fefda 100644
--- a/src/client.c
+++ b/src/client.c
@@ -1329,6 +1329,7 @@ static bool handle_client_startup(PgSocket *client, PktHdr *pkt)
 static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 {
 	SBuf *sbuf = &client->sbuf;
+	bool in_transaction = false;
 	int track_outstanding = false;
 	PreparedStatementAction ps_action = PS_IGNORE;
 	PgClosePacket close_packet;
@@ -1495,8 +1496,11 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 		client->query_start = get_cached_time();
 	}
 
+	/* xact_start is non-zero if we are inside a transaction */
+	in_transaction = client->xact_start != 0;
+
 	/* remember timestamp of the first query in a transaction */
-	if (!client->xact_start) {
+	if (!in_transaction) {
 		client->pool->stats.xact_count++;
 		client->xact_start = client->query_start;
 	}
@@ -1504,6 +1508,13 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
 
+	/* pgbouncer-rr extensions: query rewrite & client connection routing */
+	if (pkt->type == 'Q' || pkt->type == 'P') {
+		if (!rewrite_query(client, in_transaction, pkt))
+			return false;
+		route_client_connection(client, in_transaction, pkt);
+	}
+
 	/* acquire server */
 	if (!find_server(client))
 		return false;
